### PR TITLE
Add support for storage tiering and vlan enhancement

### PR DESCRIPTION
**Pull request by futuretea**
Depend on PR https://github.com/harvester/harvester/pull/2898

1. storage tiering
- [x] already support `storage_class_name` in `harvester_volume`
- [x] new field `storage_class_name` in `harvester_image`
- [x] new resources `harvester_storageclass`
2. vlan enhancement
- [x] new field `cluster_network_name` in `harvester_network`
- [x] remove fields in `harvester_clusternetwork`
- [x] new resources `harvester_vlanconfig`

**Related issues**
https://github.com/harvester/harvester/issues/2957
https://github.com/harvester/harvester/issues/2793
https://github.com/harvester/harvester/issues/2790

**Test Plan**

- Setup test env by https://github.com/harvester/terraform-provider-harvester/wiki/PR-Test-Env-Setup for `Reviwer`
- Setup test env by https://github.com/harvester/terraform-provider-harvester/wiki/Use-the-master-head-docker-image-for-testing for `Tester`
- 
- Create a test config file
```bash
vim main.tf
```
**Network**
- Add resource `harvester_clusternetwork` with `name = "mgmt"`
```hcl
resource harvester_clusternetwork mgmt {
        name = "mgmt"
}
```
```bash
terraform apply -auto-approve
```
- Expected error: ` can not create the existing mgmt clusternetwork`
- Remove resource  harvester_clusternetwork  `mgmt` from the test config file
- Add  data `harvester_clusternetwork` with `name = "mgmt"`
```hcl
data harvester_clusternetwork mgmt {
        name = "mgmt"
}
```
```bash
terraform apply -auto-approve
```
- Add resource `harvester_network` with `cluster_network_name = data.harvester_clusternetwork.mgmt`
```hcl
resource "harvester_network" "vlan1" {
  name      = "vlan1"
  namespace = "harvester-public"

  vlan_id = 1

  route_dhcp_server_ip = ""

  cluster_network_name = data.harvester_clusternetwork.mgmt.name
}
```
```bash
terraform apply -auto-approve
```
- Add `harvester_clusternetwork` `test`
```hcl
resource harvester_clusternetwork test{
        name = "test"
}
```
```bash
terraform apply -auto-approve
```
- Add resource `harvester_network` with `cluster_network_name = "test"`
```hcl
resource "harvester_network" "test-vlan1" {
  name      = "test-vlan1"
  namespace = "harvester-public"

  vlan_id = 1

  route_dhcp_server_ip = ""

  cluster_network_name = harvester_clusternetwork.test.name
}
```
```bash
terraform apply -auto-approve
```
- Expected error: `can not use the unready clusternetwork test in networks`
- Remove resource  harvester_network `test-vlan1` from the test config file
- Add resource `harvester_vlanconfig` with `cluster_network_name = "test"`
(need change uplink and node_selector for your cluster)
```hcl
resource "harvester_vlanconfig" "test" {
  name = "test"

  cluster_network_name = harvester_clusternetwork.test.name

  uplink {
    nics = [
      "ens36",
    ]

    bond_mode = "active-backup"
    mtu       = 1500
  }

  node_selector = {
    "kubernetes.io/hostname" : "harv-110rc3"
  }
}
```
```bash
terraform apply -auto-approve
```
- Add resource `harvester_network` with `cluster_network_name = "test"`
```hcl
resource "harvester_network" "test-vlan1" {
  name      = "test-vlan1"
  namespace = "harvester-public"

  vlan_id = 1

  route_dhcp_server_ip = ""

  cluster_network_name = harvester_clusternetwork.test.name
  depends_on = [
    harvester_vlanconfig.test
  ]
}
```
```bash
terraform apply -auto-approve
```

**Storage tiering**
- Add Tag `ssd` and `nvme` to one disk in Harvester cluaster via `Hosts > Edit Config`
image

- Add resource `harvester_storageclass` with `name = "test"`
```hcl
resource "harvester_storageclass" "test" {
  name = "test"

  parameters = {
    "migratable"          = "true"
    "numberOfReplicas"    = "1"
    "staleReplicaTimeout" = "30"
    "diskSelector"        = "ssd,nvme"
  }
}
```
```bash
terraform apply -auto-approve
```
- Add resource `harvester_image` with `storage_class_name = "test"`

```hcl
resource "harvester_image" "ubuntu20-ssd" {
  name      = "ubuntu20-ssd"
  namespace = "harvester-public"

  storage_class_name = harvester_storageclass.test.name

  display_name = "ubuntu20-ssd"
  source_type  = "download"
  url          = "http://cloud-images.ubuntu.com/releases/focal/release/ubuntu-20.04-server-cloudimg-amd64.img"
}
```
```bash
terraform apply -auto-approve
```
- Add resource `harvester_image` without `storage_class_name`
```hcl
resource "harvester_image" "ubuntu20" {
  name      = "ubuntu20"
  namespace = "harvester-public"

  display_name = "ubuntu20"
  source_type  = "download"
  url          = "http://cloud-images.ubuntu.com/releases/focal/release/ubuntu-20.04-server-cloudimg-amd64.img"
}
```
```bash
terraform apply -auto-approve
```

- Add resource `harvester_volume` with `storage_class_name = "test"`
```hcl
resource "harvester_volume" "ubuntu20-dev-mount-ssd-disk" {
  name      = "ubuntu20-dev-mount-ssd-disk"
  namespace = "default"

  storage_class_name = harvester_storageclass.test.name

  size = "10Gi"
}
```
```bash
terraform apply -auto-approve
```

- Add resource `harvester_volume` without `storage_class_name·
```hcl
resource "harvester_volume" "ubuntu20-dev-mount-disk" {
  name      = "ubuntu20-dev-mount-disk"
  namespace = "default"

  size = "10Gi"
}
```
```bash
terraform apply -auto-approve
```

- Add resource `harvester_virtualmachine`
```hcl
resource "harvester_virtualmachine" "ubuntu20-dev" {
  name                 = "ubuntu20-dev"
  namespace            = "default"
  restart_after_update = true

  description = "test raw image"
  tags = {
    ssh-user = "ubuntu"
  }

  cpu    = 2
  memory = "2Gi"

  efi         = true
  secure_boot = true

  run_strategy = "RerunOnFailure"
  hostname     = "ubuntu-dev"
  machine_type = "q35"

  network_interface {
    name           = "nic-1"
    network_name   = harvester_network.vlan1.id
    wait_for_lease = true
  }

  network_interface {
    name         = "nic-2"
    model        = "virtio"
    type         = "bridge"
    network_name = harvester_network.test-vlan1.id
  }

  disk {
    name       = "rootdisk"
    type       = "disk"
    size       = "10Gi"
    bus        = "virtio"
    boot_order = 1

    image       = harvester_image.ubuntu20-ssd.id
    auto_delete = true
  }

  disk {
    name        = "emptydisk"
    type        = "disk"
    size        = "20Gi"
    bus         = "virtio"
    auto_delete = true
  }

  disk {
    name = "mount-disk-ssd"
    type = "disk"
    bus  = "scsi"

    existing_volume_name = harvester_volume.ubuntu20-dev-mount-ssd-disk.name
    auto_delete          = false
    hot_plug             = true
  }

  cloudinit {
    user_data    = <<-EOF
      #cloud-config
      user: ubuntu
      password: root
      chpasswd:
        expire: false
      ssh_pwauth: true
      package_update: true
      packages:
        - qemu-guest-agent
      runcmd:
        - - systemctl
          - enable
          - '--now'
          - qemu-guest-agent
      EOF
    network_data = ""
  }
}
```

```bash
terraform apply -auto-approve
```

**Test import**
- remove local states
```bash
mv terraform.tfstate terraform.tfstate.backup
```

- import resources
```bash
terraform import harvester_clusternetwork.test test
terraform import harvester_network.vlan1 harvester-public/vlan1
terraform import harvester_vlanconfig.test test
terraform import harvester_network.test-vlan1 harvester-public/test-vlan1
terraform import harvester_storageclass.test test
terraform import harvester_image.ubuntu20-ssd harvester-public/ubuntu20-ssd
terraform import harvester_image.ubuntu20 harvester-public/ubuntu20
terraform import harvester_volume.ubuntu20-dev-mount-ssd-disk default/ubuntu20-dev-mount-ssd-disk
terraform import harvester_volume.ubuntu20-dev-mount-disk default/ubuntu20-dev-mount-disk
terraform import harvester_virtualmachine.ubuntu20-dev default/ubuntu20-dev

```
```bash
terraform show
```

**Test Destroy**
```bash
terraform destroy -auto-approve
```

-------

**The Content of main.tf**

```hcl
data harvester_clusternetwork mgmt {
        name = "mgmt"
}
resource "harvester_network" "vlan1" {
  name      = "vlan1"
  namespace = "harvester-public"

  vlan_id = 1

  route_dhcp_server_ip = ""

  cluster_network_name = data.harvester_clusternetwork.mgmt.name
}

resource harvester_clusternetwork test{
        name = "test"
}

resource "harvester_vlanconfig" "test" {
  name = "test"

  cluster_network_name = harvester_clusternetwork.test.name

  uplink {
    nics = [
      "ens36",
    ]

    bond_mode = "active-backup"
    mtu       = 1500
  }

  node_selector = {
    "kubernetes.io/hostname" : "harv-110rc3"
  }
}
resource "harvester_network" "test-vlan1" {
  name      = "test-vlan1"
  namespace = "harvester-public"

  vlan_id = 1

  route_dhcp_server_ip = ""

  cluster_network_name = harvester_clusternetwork.test.name
  depends_on = [
    harvester_vlanconfig.test
  ]
}

resource "harvester_storageclass" "test" {
  name = "test"

  parameters = {
    "migratable"          = "true"
    "numberOfReplicas"    = "1"
    "staleReplicaTimeout" = "30"
    "diskSelector"        = "ssd,nvme"
  }
}

resource "harvester_image" "ubuntu20-ssd" {
  name      = "ubuntu20-ssd"
  namespace = "harvester-public"

  storage_class_name = harvester_storageclass.test.name

  display_name = "ubuntu20-ssd"
  source_type  = "download"
  url          = "http://cloud-images.ubuntu.com/releases/focal/release/ubuntu-20.04-server-cloudimg-amd64.img"
}

resource "harvester_image" "ubuntu20" {
  name      = "ubuntu20"
  namespace = "harvester-public"

  display_name = "ubuntu20"
  source_type  = "download"
  url          = "http://cloud-images.ubuntu.com/releases/focal/release/ubuntu-20.04-server-cloudimg-amd64.img"
}

resource "harvester_volume" "ubuntu20-dev-mount-ssd-disk" {
  name      = "ubuntu20-dev-mount-ssd-disk"
  namespace = "default"

  storage_class_name = harvester_storageclass.test.name

  size = "10Gi"
}

resource "harvester_volume" "ubuntu20-dev-mount-disk" {
  name      = "ubuntu20-dev-mount-disk"
  namespace = "default"

  size = "10Gi"
}

resource "harvester_virtualmachine" "ubuntu20-dev" {
  name                 = "ubuntu20-dev"
  namespace            = "default"
  restart_after_update = true

  description = "test raw image"
  tags = {
    ssh-user = "ubuntu"
  }

  cpu    = 2
  memory = "2Gi"

  efi         = true
  secure_boot = true

  run_strategy = "RerunOnFailure"
  hostname     = "ubuntu-dev"
  machine_type = "q35"

  network_interface {
    name           = "nic-1"
    network_name   = harvester_network.vlan1.id
    wait_for_lease = true
  }

  network_interface {
    name         = "nic-2"
    model        = "virtio"
    type         = "bridge"
    network_name = harvester_network.test-vlan1.id
  }

  disk {
    name       = "rootdisk"
    type       = "disk"
    size       = "10Gi"
    bus        = "virtio"
    boot_order = 1

    image       = harvester_image.ubuntu20-ssd.id
    auto_delete = true
  }

  disk {
    name        = "emptydisk"
    type        = "disk"
    size        = "20Gi"
    bus         = "virtio"
    auto_delete = true
  }

  disk {
    name = "mount-disk-ssd"
    type = "disk"
    bus  = "scsi"

    existing_volume_name = harvester_volume.ubuntu20-dev-mount-ssd-disk.name
    auto_delete          = false
    hot_plug             = true
  }

  cloudinit {
    user_data    = <<-EOF
      #cloud-config
      user: ubuntu
      password: root
      chpasswd:
        expire: false
      ssh_pwauth: true
      package_update: true
      packages:
        - qemu-guest-agent
      runcmd:
        - - systemctl
          - enable
          - '--now'
          - qemu-guest-agent
      EOF
    network_data = ""
  }
}
```

### CODE REVIEW

File: internal/provider/clusternetwork/resource_clusternetwork_constructor.go [hunk 1]
```diff
@@ -18,25 +15,8 @@ type Constructor struct {
...
-					return fmt.Errorf("%s is true, please specify %s", constants.FieldClusterNetworkEnable, constants.FieldClusterNetworkDefaultPhysicalNIC)
-				}
-				c.ClusterNetwork.Config[constants.ClusterNetworkConfigKeyDefaultPhysicalNIC] = defaultPhysicalNIC
-				return nil
-			},
-			Required: true,
-		},
-	}
+	processors := util.NewProcessors().Tags(&c.ClusterNetwork.Labels).Description(&c.ClusterNetwork.Annotations)
+	customProcessors := []util.Processor{}
```
[the remark this answers is not in the document]
Comment (futuretea) [reply internal/provider/clusternetwork/resource_clusternetwork_constructor.go:19]: As a placeholder, just to be consistent with other resources.

**Review by masteryyh — APPROVED** [review]
LGTM

**Comment by chrisho** [comment]
LGTM

**Review by chrisho — APPROVED** [review]
LGTM

File: examples/resources/harvester_image/resource.tf [hunk 2]
```diff
@@ -14,4 +14,15 @@ resource "harvester_image" "ubuntu20" {
...
+
+resource "harvester_image" "ubuntu20-ssd" {
+  name      = "ubuntu20-ssd"
+  namespace = "harvester-public"
+
+  storage_class_name = harvester_storageclass.ssd-replicas-3.name
+
+  display_name = "ubuntu20-ssd"
+  source_type  = "download"
+  url          = "http://cloud-images.ubuntu.com/releases/focal/release/ubuntu-20.04-server-cloudimg-amd64.img"
```
[the remark this answers is not in the document]
Comment (futuretea) [reply examples/resources/harvester_image/resource.tf]: Updated

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

**Pull request merged**